From dc11453c1ac5ca3255ad94a5c18c4456ea986289 Mon Sep 17 00:00:00 2001
From: "pj.chen" <chen.pj@inventec.com>
Date: Tue, 15 Mar 2022 06:55:29 +0000
Subject: [PATCH 20/30] Bug 876 - [SW][Common][Intel-ipmi-oem] - Remove sensors
 has no sensorInfo in subtree

Symptom/Reason:
    Intel-ipmi-oem would scan all dbus object path contains sensor relate interfaces.
    But if sensor didn't contain sensorInfo, would cause SDR function error.

Root Cause:
    SDR is relate to sensorTree size, but also needs to get sensor
    number in sensorInfo.

Solution/Change:
    Remove sensor without sensorInfo in sensorTree.

Entry Test:
sysadmin@starscream:~# ipmitool sdr list
....
Without error

%% original patch: 0020-Bug-876-SW-Common-Intel-ipmi-oem-Remove-sensors-has-.patch
---
 include/sdrutils.hpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/include/sdrutils.hpp b/include/sdrutils.hpp
index c9f69bc..56fff11 100644
--- a/include/sdrutils.hpp
+++ b/include/sdrutils.hpp
@@ -287,6 +287,23 @@ inline static uint16_t getSensorSubtree(std::shared_ptr<SensorSubTree>& subtree,
         phosphor::logging::log<phosphor::logging::level::ERR>(e.what());
         return -1;
     }
+
+    /* Bug 876 - [SW][Common][Intel-ipmi-oem] - Remove sensors has no sensorInfo in subtree */
+    std::vector<std::string> removeTarget;
+    for (auto& sensor : *sensorTreePtr)
+    {
+        auto& subVect = sensor.second.begin()->second;
+
+        if (std::find(subVect.begin(), subVect.end(), ipmiInterface) == subVect.end())
+        {
+            removeTarget.push_back(sensor.first);
+        }
+    }
+    for(auto& it : removeTarget)
+    {
+        sensorTreePtr->erase(it);
+    }
+
     subtree = sensorTreePtr;
     sensorUpdatedIndex++;
     // The SDR is being regenerated, wipe the old stats
-- 
2.34.1

